pubsub/cloud-client/quickstart/sub_test.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python
+
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import mock
+import os
+import pytest
+import time
+
+from google.cloud import pubsub_v1
+
+import sub
+
+
+PROJECT = os.environ['GCLOUD_PROJECT']
+TOPIC = 'quickstart-sub-test-topic'
+SUBSCRIPTION = 'quickstart-sub-test-topic-sub'
+
+
+@pytest.fixture(scope='module')
+def publisher_client():
+    yield pubsub_v1.PublisherClient()
+
+
+@pytest.fixture(scope='module')
+def topic(publisher_client):
+    topic_path = publisher_client.topic_path(PROJECT, TOPIC)
+
+    try:
+        publisher_client.delete_topic(topic_path)
+    except Exception:
+        pass
+
+    publisher_client.create_topic(topic_path)
+
+    yield topic_path
+
+
+@pytest.fixture(scope='module')
+def subscriber_client():
+    yield pubsub_v1.SubscriberClient()
+
+
+@pytest.fixture(scope='module')
+def subscription(subscriber_client, topic):
+    subscription_path = subscriber_client.subscription_path(
+        PROJECT, SUBSCRIPTION)
+
+    try:
+        subscriber_client.delete_subscription(subscription_path)
+    except Exception:
+        pass
+
+    subscriber_client.create_subscription(subscription_path, topic=topic)
+
+    yield SUBSCRIPTION
+
+
+def _publish_messages(publisher_client, topic):
+    data = u'Hello, World!'.encode('utf-8')
+    publisher_client.publish(topic, data=data)
+
+
+def _make_sleep_patch():
```

Comment:
Prefer using the `monkeypatch` fixture for pytest. It can clean up after your tests so that the sleep function is restored after the test is done.

Reply:
I need some help with this. In the meanwhile, since other tests are also written using `_make_sleep_patch` in the same repo, shall we start another PR just for this purpose?

Reply:
This is the pattern I'm describing: https://stackoverflow.com/a/29110609/101923,

Now that I have a closer look, `mock.patch` does the same thing, but this still seems much more complicated than it needs to be.

Reply:
I tried `monkeypatch` but it created an infinite loop. 

```python
def mock_sleep(sleep):
    time.sleep(10)
    raise RuntimeError('sigil')

monkeypatch.setattr(time, 'sleep', mock_sleep)
```

`while True: time.sleep(60)` would call `time.sleep(10)`, which calls `time.sleep(10)` and never reaches `raise RuntimeError('sigil')`.